### DISCUSSION TITLE
Export Rust shims and bindings

### DISCUSSION THREAD

**Discussion by ten3roberts** [wasm-bindgen/wasm-bindgen#3499] [Q&A] [answered]
### Summary

A clear and concise summary of your question.

`wasm-bindgen` cli can consume a compiled `.wasm` file and through the use of the `__wbindgen_describe` functions get reflection data and generate appropriate wrapper functions in javascript that will allow passings strings and js-objects.

Is there a method to get the same functionality but generate the bindings for rust functions?

### Additional Details

The use case I am having is a wasm module which downloads another wasm file over the network.

The binary and url is not known, but the API of them are the same.

The wasm file then needs to call functions of the downloaded wasm using complex types such as strings.

Additionally, the downloaded client wasm makes use of wasm-bindgen internally and links to `web-sys` et alii.

I am aware that calling an "FFI style" function from the parent wasm is possible, but it is not feasable to uphold the invariants that the wasm_bindgen call marshalling expects, as well as the format not being stable or necessarily public.

Including the wasm file as a rust dependency and linking it in is not possible.

Thanks for your input

**Comment by daxpedda** [comment, marked as answer]
> Is there a method to get the same functionality but generate the bindings for rust functions?

`wasm-bindgen` doesn't offer this functionality, no.
> Additionally, the downloaded client wasm makes use of wasm-bindgen internally and links to `web-sys` et alii.

If this Wasm file is using `wasm-bindgen`, it won't work properly if you don't run it through `wasm-bindgen` and don't have it's JS shim anyway. How do you intend to solve that in the first place?

---

I'm assuming you want to use Wasm files provided by a client and use them in your Wasm application that is running in a browser.
Unfortunately there isn't an easy way to do this right now, but what you need is an FFI interface that both you and the client agree on beforehand. Something that `wasm-bindgen` doesn't provide.

I recommend you to look into [`wit-bindgen`](https://crates.io/crates/wit-bindgen).

**Reply by ten3roberts** [reply]
Thank you. I was experimenting more with this and realized that apart from the shim you aslo need to set up the WebAssembly imports object to bind the "linked" functions.

I added `Date::now` to the client wasm and that means I have to also provide a field `__wbg_now_...` in the imports object and so on.

So it seems that you can not just instantiate a wasm-bindgen dependent wasm, it is in a way a filetype or domain of its own.

I do think admitting defeat in the Rust stack and shelling out to JS for this (and in doing so fetching two files) to enter the client wasm.

I will be using `no-modules` for this approach, as it seems to be able to have wasm data provided to it rather than fetching it itself, which works well for a separately hosted file server.

Thanks for the help.
